### ISSUE TITLE
WAL file explodes in size during ingestion if running any concurrent reads/writes queries

### ISSUE THREAD

**Issue by k-anshul** [duckdb/duckdb#9150]
### What happens?

While duckdb is ingesting some data if any read/write is triggered using another connection then `.db.wal` file intermittently explodes in size. Adding a python script that reproduces this issue with data. Please note this is not specific to this data and occurs for other datasets as well. Also reproduced with go code which uses C APIs internally.
I would also like to point few observations:
1. This only happens when running queries that use data stored on disk and nothing happens if running queries like `select 1`
2. For bigger datasets the wal file can be as big as 10 times as that of actual db file (example: db file : 50GB and `wal` file 500 GB) and duckdb spends huge time in clearing the wal file. This is usually observed if running a `rename table query`.

### To Reproduce

1. Run the python script in this repo : https://github.com/k-anshul/wal_debug
2. While the script is running observe that WAL file intermittently explodes in size (>1GB)

### OS:

mac os aarch64, linux amd64

### DuckDB Version:

0.8.1,0.9.0

### DuckDB Client:

Python

### Full Name:

Anshul Khandelwal

### Affiliation:

Rill Data

### Have you tried this on the latest `main` branch?

I have tested with a main build

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

**Comment by Mytherin** [comment]
Thanks for the report! This is currently an expected restriction of the way that checkpointing works that we hope to alleviate in the future. The database cannot issue a checkpoint command if there are concurrent transactions running. You can run a `FORCE CHECKPOINT` command to block all concurrent connections and force a checkpoint (flushing the WAL into the persistent storage).

**Comment by begelundmuller** [comment]
@Mytherin, thanks for looking into this! We experimented with `FORCE CHECKPOINT`, is it expected that it leads to concurrent queries erroring? Also, is there a way to make DuckDB use force checkpoints during long-running queries like renames or `CREATE TABLE ... AS SELECT ...` queries to limit the `.wal` size?

Taking a step back, do you have any advice on how we can safely reload a table in the background while serving queries? Over the weekend, we saw a crash where the `.db` file had grown to 199GB and the `.wal` file had grown to 1.7TB. This is for the same dataset, for which the `.db` file is just 54GB when loaded for the first time.

This is the sequence of steps we currently run to reload a table:
1. Download the source data in batches of 2GB
2. Create the first batch as a staging table – `CREATE TABLE xxx_tmp AS SELECT * FROM read_parquet({files})`
3. Incrementally insert batches into the staging table – `INSERT INTO xxx_tmp AS SELECT * FROM read_parquet({files})`
4. Transactionally replace the current table with the staging table – `DROP TABLE xxx; ALTER TABLE xxx_tmp RENAME TO xxx;`

We are currently investigating a different approach where we:
1. Load each table in a new `xxx_[version].db` file
2. Run `ATTACH 'xxx_[version].db'` on the main DB handle
3. Run `CREATE OR REPLACE VIEW xxx AS SELECT * FROM xxx_[version].xxx` to replace the old table with a newly ingested one
4. Delete the `xxx_[version].db` file for the old table that is no longer used

This seems to keep `.wal` and `.db` sizes normal. But I'm wondering if there's a simpler approach or if this might lead to other issues down the line?

**Comment by Mytherin** [comment]
> @Mytherin, thanks for looking into this! We experimented with FORCE CHECKPOINT, is it expected that it leads to concurrent queries erroring?

Oh yes, you are right, it actually cancels all concurrently running queries in order to force the checkpoint.

> Taking a step back, do you have any advice on how we can safely reload a table in the background while serving queries?

Currently any concurrently running queries will block checkpointing, so if you never have gaps and connections are always active reading data then the database will never be checkpointed and the WAL will grow without bounds. As you are constantly ingesting a lot of data into new tables this means the WAL will grow very fast as well.

Attaching and detaching separate databases seems like a good work-around for now. I can't think of any immediate issues you would run into with this model.

**Comment by begelundmuller** [comment]
> so if you never have gaps and connections are always active reading data then the database will never be checkpointed and the WAL will grow without bounds

This makes me wonder if there could be another issue at play here – while we have multiple connections open and concurrent queries may occur, there are also long periods of time (minutes or hours) without any queries (though the connections stay open in a pool).

We are also not really ingesting data constantly – for the case of the ~50 GB dataset that turned into a 1.7TB `.wal` over the course of a few days, it was just refreshed once per day (when successful, it takes approx. 15 minutes for a refresh, including download times).

Does that still seem congruent with your expectations or could something else be wrong?

> Attaching and detaching separate databases seems like a good work-around for now. I can't think of any immediate issues you would run into with this model.

Thanks, we're moving ahead with this workaround for now :)

**Comment by Mytherin** [comment]
> We are also not really ingesting data constantly – for the case of the ~50 GB dataset that turned into a 1.7TB .wal over the course of a few days, it was just refreshed once per day (when successful, it takes approx. 15 minutes for a refresh, including download times).

The WAL stores raw data (not compressed data) so this likely introduces significant bloat. A 50GB compressed dataset might be 300GB uncompressed. The WAL is in general not intended for storage of large amounts of data but rather for smaller changes to the database.

**Comment by msf** [comment]
Hello there.
We're also having this issue in production where the WAL grows unbounded as long as there is activity in the server (so concurrent reads & writes are ongoing).
We (Dune Analytics) have a repro using go-duckdb and concurrent read and write activity. This repro leverages our abstractions, so we need to strip it down to file a good repro example.

**Comment by XBeg9** [comment]
same here, I am ingesting quite a large dataset (800 files of jsonl each 1gb chunk) using `read_json_auto('/mnt/data2/jobs/chunk*.jsonl', union_by_name=true, format='newline_delimited')` and .wal blow into equal-size

**Comment by Mytherin** [comment]
This is indeed still an expected restriction - checkpoints are still blocked by concurrent read and write transactions, meaning the WAL file will grow until either a `FORCE CHECKPOINT` is run, or until there is a transaction that finishes without concurrent transactions. The previous PR only fixed this partially.
